apps/notifications/src/panel/state/create-listener-middleware.ts

```diff
@@ -0,0 +1,90 @@
+import type { MiddlewareAPI, AnyAction, Action, Middleware } from 'redux';
+
+type Handler = ( storeAPI: MiddlewareAPI, action: Action ) => void;
+
+interface AddListenersAction extends Action {
+	type: 'LISTENER-MIDDLEWARE/ADD_LISTENERS';
+	payload: Record< string, Handler[] >;
+}
+interface RemoveListenersAction extends Action {
+	type: 'LISTENER-MIDDLEWARE/REMOVE_LISTENERS';
+	payload: Record< string, Handler[] >;
+}
+interface ClearListenersAction extends Action {
+	type: 'LISTENER-MIDDLEWARE/CLEAR_LISTENERS';
+}
+type MiddlewareAction = AddListenersAction | RemoveListenersAction | ClearListenersAction;
+
+function createListenerMiddleware(): Middleware {
+	const listeners = new Map< string, Set< Handler > >();
+
+	return ( storeAPI: MiddlewareAPI ) =>
+		( next: ( action: Action ) => void ) =>
+		( anyAction: AnyAction ) => {
+			// We only want to notify listeners of actions that are objects.
+			// (That is, ignore thunks and other non-action objects.)
+			if ( typeof anyAction !== 'object' || typeof anyAction.type !== 'string' ) {
+				return next( anyAction );
+			}
```

Comment:
This check is superfluous if you make sure that the listener middleware always runs after the thunk middleware. Which it does, because it's registered later.

Reply:
I think it's a good practice to keep this independent from the setup, especially given that this is not a costly operation, WDYT?

Reply:
Yes, it's not a big deal, it's true that all Redux middlewares have a condition like "if the action is not my kind of thing, pass it throught to `next` immediately".

What's different here is that if the listener middleware encounters a non-plain action, it means that there is a bug: it's supposed to intercept all plain actions, but if registered in wrong order, it will miss the plain actions generated by other middlewares, like `redux-thunk`.

But it's no big deal, feel free to keep it as it is.